endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/ArgumentValues.kt

```diff
@@ -0,0 +1,178 @@
+     */
+    protected abstract fun saveValue(stringValue: String)
+
+    /**
+     * Set value of delegated property.
+     */
+    fun setDelegatedValue(providedValue: TResult) {
+        parsedValue = providedValue
+        valueOrigin = ArgParser.ValueOrigin.REDEFINED
+    }
+
+    /**
+     * Add parsed value from command line.
+     *
+     * @param stringValue value from command line.
+     */
+    fun addValue(stringValue: String) {
+        // Check of possibility to set several values to one option/argument.
+        if (descriptor is OptionDescriptor<*, *> && !descriptor.multiple &&
+                !isEmpty() && descriptor.delimiter == null) {
+            throw ParsingException("Try to provide more than one value for ${descriptor.fullName}.")
+        }
+        // Show deprecated warning only first time of using option/argument.
+        descriptor.deprecatedWarning?.let {
+            if (isEmpty())
+                println ("Warning: $it")
+        }
+        // Split value if needed.
+        if (descriptor is OptionDescriptor<*, *> && descriptor.delimiter != null) {
+            stringValue.split(descriptor.delimiter).forEach {
+                saveValue(it)
+            }
+        } else {
+            saveValue(stringValue)
+        }
+    }
+
+    /**
+     * Set default value to option.
+     */
+    fun addDefaultValue() {
+        if (!descriptor.defaultValueSet && descriptor.required) {
+            throw ParsingException("Please, provide value for ${descriptor.textDescription}. It should be always set.")
+        }
+        if (descriptor.defaultValueSet) {
+            parsedValue = descriptor.defaultValue!!
+            valueOrigin = ArgParser.ValueOrigin.SET_DEFAULT_VALUE
+        }
+    }
+
+    /**
+     * Provide name for CLI entity.
+     *
+     * @param name name for CLI entity.
+     */
+    fun provideName(name: String) {
+        descriptor.fullName ?: run { descriptor.fullName = name }
+    }
+}
+
+/**
+ * Single argument value.
+ *
+ * @property descriptor descriptor of option/argument.
+ */
+internal abstract class AbstractArgumentSingleValue<T: Any>(descriptor: Descriptor<T, T>):
+        ParsingValue<T, T>(descriptor) {
+
+    override fun saveValue(stringValue: String) {
+        if (!valueIsInitialized()) {
+            parsedValue = descriptor.type.conversion(stringValue, descriptor.fullName!!)
+            valueOrigin = ArgParser.ValueOrigin.SET_BY_USER
+        } else {
+            throw ParsingException("Try to provide more than one value $parsedValue and $stringValue for ${descriptor.fullName}.")
+        }
+    }
+
+    override fun isEmpty(): Boolean = !valueIsInitialized()
+}
+
+/**
+ * Single argument value.
+ *
+ * @property descriptor descriptor of option/argument.
+ */
+internal class ArgumentSingleValue<T: Any>(descriptor: Descriptor<T, T>): AbstractArgumentSingleValue<T>(descriptor),
+        ArgumentValueDelegate<T> {
+    override val value: T
```

Comment:
Created field to provide opportunity to get value if I don't use delegated properties.